compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java

```diff
@@ -288,79 +289,116 @@ public BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType ta
-            case TypeTags.XML_COMMENT:
-            case TypeTags.XML_ELEMENT:
-            case TypeTags.XML_PI:
-            case TypeTags.XML_TEXT:
-                selectType = checkExpr(selectExp, env, type, data);
-                resolvedType = selectType;
-                break;
-            case TypeTags.INTERSECTION:
-                type = ((BIntersectionType) type).effectiveType;
-                solveSelectTypeAndResolveType(queryExpr, selectExp, type, collectionType, selectTypes,
-                        resolvedTypes, env, data, Symbols.isFlagOn(type.flags, Flags.READONLY));
-                return;
-            case TypeTags.NONE:
-            default:
-                // contextually expected type not given (i.e var).
-                selectType = checkExprSilent(nodeCloner.cloneNode(selectExp), type, data);
-                if (selectType != symTable.semanticError) {
-                    selectType = checkExpr(selectExp, env, type, data);
-                }  else {
-                    selectType = checkExpr(selectExp, env, data);
+        List<BType> possibleSelectTypes = new ArrayList<>();
+        List<BType> possibleResolvedTypes = new ArrayList<>();
+        LinkedHashSet<BType> errorTypes = new LinkedHashSet<>();
+        for (BType expType : expTypes) {
+            BType selectType, resolvedType;
+            BType type = Types.getReferredType(expType);
+            switch (type.tag) {
+                case TypeTags.ARRAY:
+                    BType elementType = ((BArrayType) type).eType;
+                    selectType = checkExprSilent(selectExp, env, elementType, data);
+                    if (selectType.tag == TypeTags.SEMANTIC_ERROR) {
+                        errorTypes.add(elementType);
+                        continue;
+                    }
+                    BType queryResultType = new BArrayType(selectType);
+                    resolvedType = getResolvedType(queryResultType, type, isReadonly, env);
+                    break;
+                case TypeTags.TABLE:
+                    BType tableConstraint = types.getSafeType(((BTableType) type).constraint,
+                            true, true);
+                    selectType = checkExprSilent(selectExp, env, tableConstraint, data);
+                    if (selectType.tag == TypeTags.SEMANTIC_ERROR) {
+                        errorTypes.add(tableConstraint);
+                        continue;
+                    }
+                    resolvedType = getResolvedType(symTable.tableType, type, isReadonly, env);
+                    break;
+                case TypeTags.STREAM:
+                    BType streamConstraint = types.getSafeType(((BStreamType) type).constraint,
+                            true, true);
+                    selectType = checkExprSilent(selectExp, env, streamConstraint, data);
+                    if (selectType.tag == TypeTags.SEMANTIC_ERROR) {
+                        errorTypes.add(selectType);
```

Comment:
Here, `streamConstraint` should be added if the `checkExpr` returns an erroneous type

Reply:
Add a test to cover each of this erronous type

Reply:
Added the `streamConstraint` to errorTypes to perform the `checkExpr` with test cases in https://github.com/ballerina-platform/ballerina-lang/pull/40271/commits/247e2b02e54ee8b97bef088ef2f9a38d18d88ba7.